test/config-luatest/box_cfg_env_test.lua

```diff
@@ -0,0 +1,105 @@
+local fun = require('fun')
+local yaml = require('yaml')
+local net_box = require('net.box')
+local t = require('luatest')
+local treegen = require('test.treegen')
+local server = require('test.luatest_helpers.server')
+local helpers = require('test.config-luatest.helpers')
+
+local g = helpers.group()
+
+local function setup(g, dir, config, env, opts)
+    local config_file = treegen.write_script(dir, 'config.yaml',
+                                             yaml.encode(config))
+    local opts = fun.chain({
+        config_file = config_file,
+        chdir = dir,
+        env = env,
+    }, opts):tomap()
+    g.server = server:new(fun.chain(opts, {alias = 'instance-001'}):tomap())
+    g.server:start()
+end
+
+local function reload(g)
+    g.server:exec(function()
+        local config = require('config')
+
+        config:reload()
+    end)
+    g.server:wait_until_ready()
+end
+
+-- Verify that TT_LISTEN works before and after config:reload().
+g.test_basic = function(g)
+    -- Prepare a config w/o iproto.listen.
+    local config = table.copy(helpers.simple_config)
+    config.iproto = nil
+
+    -- Start the server.
+    local dir = treegen.prepare_directory(g, {}, {})
+    local net_box_uri = ('unix/:%s/var/run/%s/tarantool.iproto'):format(dir,
+        'instance-001')
+    setup(g, dir, config, {
+        ['TT_LISTEN'] =
+            ('unix/:./var/run/%s/tarantool.iproto'):format('instance-001'),
+    }, {
+        net_box_uri = net_box_uri,
+    })
+
+    -- Reload and wait for its effect.
+    g.server:exec(function()
+        local config = require('config')
+
+        config:reload()
+    end)
+    g.server:wait_until_ready()
+
+    -- Just do some request.
+    local c = net_box.connect(net_box_uri)
+    local info = c:call('box.info')
+    t.assert_equals(info.id, 1)
+end
+
+-- Verify that TT_IPROTO has lower priority than cluster config's
```

Comment:
No TT_REPLICATION test.

Let's also add tests for a string option, a numeric option, a test with JSON object and JSON array in an environment variable.

Reply:
> No TT_REPLICATION test.

It is irrelevant due to #9541.

Reply:
Added more test cases.